File: internal/sources/shared/state.go
package shared

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"

	"github.com/pcr-developers/cli/internal/config"
)

// FileState tracks how many lines have been processed per file.
type FileState struct {
	mu       sync.Mutex
	data     map[string]int // file path → lines processed
	filePath string
}

func NewFileState(name string) *FileState {
	home, _ := os.UserHomeDir()
	return &FileState{
		data:     map[string]int{},
		filePath: filepath.Join(home, config.PCRDir, name+"-state.json"),
	}
}

func (s *FileState) Load() {
	s.mu.Lock()
	defer s.mu.Unlock()
	data, err := os.ReadFile(s.filePath)
	if err != nil {
		return
	}
	_ = json.Unmarshal(data, &s.data)
}

func (s *FileState) Get(path string) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.data[path]
}

func (s *FileState) Set(path string, lines int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.data[path] = lines
	s.persist()
}

func (s *FileState) persist() {
	data, err := json.MarshalIndent(s.data, "", "  ")
	if err != nil {
		return
	}
	_ = os.MkdirAll(filepath.Dir(s.filePath), 0755)
	_ = os.WriteFile(s.filePath, data, 0644)
}

File: crates/pcr-core/src/sources/shared/state.rs
//! Per-source file state tracker. Mirrors
//! `cli/internal/sources/shared/state.go` including the JSON format used
//! for persistence so an upgrade from the Go build reuses the same state.

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

use crate::config;

#[derive(Debug, Clone)]
pub struct FileState {
    inner: Arc<Mutex<FileStateInner>>,
}

#[derive(Debug)]
struct FileStateInner {
    data: HashMap<String, i64>,
    file_path: PathBuf,
}

impl FileState {
    /// Build a per-source line-state tracker rooted at
    /// `$HOME/.pcr-dev/<name>-state.json`. Panics with a clear
    /// message if `$HOME` can't be resolved — without a stable
    /// state file the watcher would silently drop its cursor on
    /// reboot (the audit's correctness concern) and re-emit every
    /// prompt on the next start. Failing fast at watcher
    /// construction is strictly better than the previous silent
    /// `/tmp` fallback. Source-watcher entry points
    /// (`vscode::watcher::run`, `claudecode::watcher::run`) are
    /// the only callers; all of them are themselves invoked from
    /// `pcr start` which has already validated the directory via
    /// `pid_file_path()?`.
    pub fn new(name: &str) -> Self {
        let file_path = config::pcr_dir()
            .expect(
                "pcr: cannot determine $HOME — refusing to put the watcher state file \
                 under /tmp (would reset capture cursors on every reboot)",
            )
            .join(format!("{name}-state.json"));
        let state = Self {
            inner: Arc::new(Mutex::new(FileStateInner {
                data: HashMap::new(),
                file_path,
            })),
        };
        state.load();
        state
    }

    pub fn load(&self) {
        let Ok(mut guard) = self.inner.lock() else {
            return;
        };
        let Ok(bytes) = std::fs::read(&guard.file_path) else {
            return;
        };
        if let Ok(loaded) = serde_json::from_slice::<HashMap<String, i64>>(&bytes) {
            guard.data = loaded;
        }
    }

    pub fn get(&self, path: &str) -> i64 {
        self.inner
            .lock()
            .ok()
            .and_then(|g| g.data.get(path).copied())
            .unwrap_or(0)
    }

    pub fn set(&self, path: &str, lines: i64) {
        let Ok(mut guard) = self.inner.lock() else {
            return;
        };
        guard.data.insert(path.to_string(), lines);
        let data = match serde_json::to_vec_pretty(&guard.data) {
            Ok(v) => v,
            Err(_) => return,
        };
        if let Some(parent) = guard.file_path.parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        let _ = std::fs::write(&guard.file_path, data);
    }
}
